packages/vite/src/node/plugins/worker.ts

```diff
@@ -276,11 +284,29 @@ export function webWorkerPlugin(config: ResolvedConfig): Plugin {
         }`,
         map: { mappings: '' } // Empty sourcemap to supress Rolup warning
       }
-    },
+    }
+  }
+}
 
-    renderChunk(code) {
-      if (isWorker && code.includes('import.meta.url')) {
-        return code.replace('import.meta.url', 'self.location.href')
+// just run in build mode
+export function webWorkerImportMetaUrlPlugin(config: ResolvedConfig): Plugin {
+  const isWorker = config.isWorker
+
+  return {
+    name: 'vite:workerImportMetaUrl',
+    transform(code, id) {
+      if (isWorker) {
+        // if build with iife it will polyfill with `rollup` will be used document in the worker
+        // else if build with es it will replace with `esbuild` will got a unexpected data.
+        // so replace import.meta.url break the default handle.
+        // And it must be done at this (after `(?new Worker)new URL('xxx', import.meta.url)` match),
+        // Otherwise, the following regex will be incorrectly matched.
+        const s = new MagicString(code, {})
```

Comment:
Could we keep this in `webWorkerPlugin` checking for `isBuild` to avoid creating another `MagicString` here? Do we really need it to be in a separate plugin?

Reply:
I can't think of a way to avoid it.

It has to come after `workerImportMetaUrlPlugin` because it would break the matching character if it was too early.

It should be possible to use `new URL('asset.js|asset.wasm', import.meta.url)` to import resources into the worker.
so it also can't between `workerPlugin` and `assetImportMetaUrlPlugin`.

I'm considering whether to put this plugin further down so that users get `import.meta.url` string.

Reply:
And why not add it to the `workerImportMetaUrlPlugin` then?

Reply:
`workerImportMetaUrlPlugin` is before `assetImportMetaUrlPlugin`

Reply:
Ok, we can leave it like this now. But I wonder if we should extract all the bits that depend on each other to a more general `importMetaUrlPlugin` so we can do all the modifications in one place.

Reply:
may be do it in `assetImportMetaUrlPlugin` but user plugin can not got a `import.meta.url` string. (they get `self.location.href`)

Reply:
We could keep it as is, for now, we'll need to review the current handling of `import.meta.url` as part of https://github.com/vitejs/vite/pull/7644 for 3.0, which will upgrade the modern target to browsers that support `import.meta.url`. So we shouldn't need to transpile to `location` anymore, but rewire the path to be relative to the bundle instead of the file IIUC

Reply:
Do you mean #7706 can keep the handling of `self.location` and `import.meta.url` handle in #7644.